### ISSUE TITLE
Repeated library panels showing only after refresh, not on first load

### ISSUE THREAD

**Issue by oliverrahner** [grafana/grafana#59805]
**What happened**:
When I open a dashboard that
- contains a libarary panel that is
- set to repeat on a variable

only the first instance of the dashboard is displayed.
When I manually click the refresh button inside Grafana (_not_ of the browser), the second instance appears.

In this GIF, the left panel is a non-library panel, the right panel is a libarary panel, both with mostly the same queries and settings behind them:

https://user-images.githubusercontent.com/2277681/205628932-dff2d770-57a2-44bc-bdfd-98c91efd5237.mp4

When I unlink that panel, it works expected immediately.
When I replace it by a (newly created) library panel, it stops working again.

**What you expected to happen**:
The second instance of the repeated panel should display on first load.

**How to reproduce it (as minimally and precisely as possible)**:

Working as expected (two panels with title "0" and "1" are showing immediately):

````json
{
  "__inputs": [
    {
      "name": "DS_INFLUXDB",
      "label": "InfluxDB",
      "description": "",
      "type": "datasource",
      "pluginId": "influxdb",
      "pluginName": "InfluxDB"
    }
  ],
  "__elements": {},
  "__requires": [
    {
      "type": "panel",
      "id": "gauge",
      "name": "Gauge",
      "version": ""
    },
    {
      "type": "grafana",
      "id": "grafana",
      "name": "Grafana",
      "version": "9.3.1"
    },
    {
      "type": "datasource",
      "id": "influxdb",
      "name": "InfluxDB",
      "version": "1.0.0"
    }
  ],
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": {
          "type": "grafana",
          "uid": "-- Grafana --"
        },
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "target": {
          "limit": 100,
          "matchAny": false,
          "tags": [],
          "type": "dashboard"
        },
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": null,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "datasource": {
        "type": "influxdb",
        "uid": "${DS_INFLUXDB}"
      },
      "description": "",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "options": {
        "orientation": "auto",
        "reduceOptions": {
          "calcs": [
            "lastNotNull"
          ],
          "fields": "",
          "values": false
        },
        "showThresholdLabels": false,
        "showThresholdMarkers": true
      },
      "pluginVersion": "9.3.1",
      "repeat": "query0",
      "repeatDirection": "v",
      "targets": [
        {
          "datasource": {
            "type": "influxdb",
            "uid": "${DS_INFLUXDB}"
          },
          "query": "",
          "refId": "A"
        }
      ],
      "title": "$query0",
      "type": "gauge"
    }
  ],
  "schemaVersion": 37,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": [
      {
        "current": {
          "selected": true,
          "text": [
            "0",
            "1"
          ],
          "value": [
            "0",
            "1"
          ]
        },
        "hide": 0,
        "includeAll": false,
        "multi": true,
        "name": "query0",
        "options": [
          {
            "selected": true,
            "text": "0",
            "value": "0"
          },
          {
            "selected": true,
            "text": "1",
            "value": "1"
          }
        ],
        "query": "0,1",
        "queryValue": "",
        "skipUrlSync": false,
        "type": "custom"
      }
    ]
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "New dashboard",
  "uid": "PNF4_aKVk",
  "version": 9,
  "weekStart": ""
}
````

Not working as expected (even refresh doesn't work, you have to unselect one of the two values and reselect it again to get both showing):

````json
{
  "__inputs": [],
  "__elements": {
    "FeB1_aFVk": {
      "name": "$query0 v2",
      "uid": "FeB1_aFVk",
      "kind": 1,
      "model": {
        "type": "gauge",
        "title": "$query0",
        "scopedVars": {
          "query0": {
            "selected": true,
            "text": "0",
            "value": "0"
          }
        },
        "repeat": "query0",
        "repeatDirection": "v",
        "datasource": {
          "type": "influxdb",
          "uid": "-b7Thb27k"
        },
        "pluginVersion": "9.3.1",
        "description": "",
        "fieldConfig": {
          "defaults": {
            "color": {
              "mode": "thresholds"
            },
            "mappings": [],
            "thresholds": {
              "mode": "absolute",
              "steps": [
                {
                  "color": "green",
                  "value": null
                },
                {
                  "color": "red",
                  "value": 80
                }
              ]
            }
          },
          "overrides": []
        },
        "options": {
          "orientation": "auto",
          "reduceOptions": {
            "calcs": [
              "lastNotNull"
            ],
            "fields": "",
            "values": false
          },
          "showThresholdLabels": false,
          "showThresholdMarkers": true
        },
        "targets": [
          {
            "datasource": {
              "type": "influxdb",
              "uid": "-b7Thb27k"
            },
            "query": "",
            "refId": "A"
          }
        ]
      }
    }
  },
  "__requires": [
    {
      "type": "grafana",
      "id": "grafana",
      "name": "Grafana",
      "version": "9.3.1"
    }
  ],
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": {
          "type": "grafana",
          "uid": "-- Grafana --"
        },
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "target": {
          "limit": 100,
          "matchAny": false,
          "tags": [],
          "type": "dashboard"
        },
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": null,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "libraryPanel": {
        "uid": "FeB1_aFVk",
        "name": "$query0 v2"
      }
    }
  ],
  "schemaVersion": 37,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": [
      {
        "current": {
          "selected": true,
          "text": [
            "0",
            "1"
          ],
          "value": [
            "0",
            "1"
          ]
        },
        "hide": 0,
        "includeAll": false,
        "multi": true,
        "name": "query0",
        "options": [
          {
            "selected": true,
            "text": "0",
            "value": "0"
          },
          {
            "selected": true,
            "text": "1",
            "value": "1"
          }
        ],
        "query": "0,1",
        "queryValue": "",
        "skipUrlSync": false,
        "type": "custom"
      }
    ]
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "New dashboard",
  "uid": "PNF4_aKVk",
  "version": 8,
  "weekStart": ""
}
````

**Anything else we need to know?**:
Crocodiles cannot stick their tongue out.

**Environment**:
- Grafana version: 9.3.1 (updated from 8.5.2, worked there)
- Data source type & version: InfluxDB 2.1
- OS Grafana is installed on: Ubuntu 20.04/Docker
- User OS & Browser: Win 11, Chrome & Edge; FullPageOS, Chrome
- Grafana plugins: -
- Others:

**Comment by zuchka** [comment]
Thanks for creating this issue, @oliverrahner 

troubleshooting request: can you try on 9.2.6? I see this in the changelog for 9.2.7:

**LibraryPanels:** Load library panels in the frontend rather than the backend. [#50560](https://github.com/grafana/grafana/pull/50560), [@ryantxu](https://github.com/ryantxu)

**Comment by oliverrahner** [comment]
Will do this tomorrow!

**Comment by oliverrahner** [comment]
Did I mention that my personal days have 2^8 hours for efficiency reasons? 😅

I tried in 9.2.6 and 9.2.7, both don't seem to show the reported behavior.
It starts to break on 9.3.0.

**Comment by ivanortegaalba** [comment]
We could reproduce it quickly in [this dashboard](https://play.grafana.org/d/c8NKNK5Vz/repeating-library-panels-dashboard-copy-4634?var-query0=All&orgId=1). I can confirm this doesn't happen with non-library panels.

Quick steps to reproduce:
1. Open [this dashboard](https://play.grafana.org/d/c8NKNK5Vz/repeating-library-panels-dashboard-copy-4634?var-query0=All&orgId=1)
2. There is only 1 chart displayed, but it should be 2 instead
3. Select the two options manually and two charts will appear
4. Refresh the page, the chart is not repeated again

**Comment by Rumbelstilzchen** [comment]
My usecase is a little bit different but the error seems to be related.
I use "add new row" and configure it with title: ${varname:raw} ;repeat for: varname
Within the query of the library panel again the ${varname:raw} is used.

  * The repetition and the labelling of each row is always correct, no matter whether single value or 'all' is used
  * with query-inspector of any library panel I can see that the ${varname:raw} is resolved to be always the same
  
  --> label for row and data are no longer in sync
So far I did not figure out by witch logic the  ${varname:raw} of the library panel is resolved to, but it is the same value for all rows.

Is this connected? Shall I open a new issue?
Versions affected: 9.3.1 & 9.3.2

**Comment by Rumbelstilzchen** [comment]
After removing
```json
"scopedVars": {
  "query0": {
    "selected": true,
    "text": "0",
    "value": "0"
  }
},
```
from the json of all Library-Panels it seems to work as expected. (Removal was done directly within sqlite)
I cannot check for the other mentioned issue here as I do not have access to the underlying json/grafana.db.

As long as the panel is not edited, this is not added again.

**Comment by kaydelaney** [comment]
So I believe the problem here is that prior to [this PR](https://github.com/grafana/grafana/pull/50560), library panels were loaded in the backend, and so information about that panel's repeats was included with the initial dashboard payload, letting the dashboard display the repeats as expected.
The linked PR changed things to instead load the library panels on the frontend, so rather than having the initial dashboard saturated with the library panel models before being returned, we just get the library panel ID, which is used to fetch the library panel when it gets lazy-loaded (i.e. scrolled into view). We don't currently process repeats when a panel is loaded, so the dashboard model's `processRepeats` method doesn't get called.

I think there are two solutions we could go for:
1. Call `processRepeats` after a library panel gets scrolled into view and loaded. This would be the simplest to implement I think but could introduce some layout jank from the repeated panels.
2. Saturate the dashboard model with the library panel's repeat info before it gets returned. This could reduce the layout jank mentioned earlier, but it would mean reintroducing some of the code that was removed in the PR mentioned, increasing the backend's complexity.

Any thoughts @ryantxu ?

**Comment by midN** [comment]
👍 On the bug, are there any updates on the issue by chance @kaydelaney @natellium ?
Looks like the issue was in progress but then was downgraded into backlog.

We are running into same issue where on initial load some panels are randomly failing to load with no obvious error, switching variables gets it to work again.

**Comment by natellium** [comment]
Hey @midN, sorry for the wait. I will bring it to the team to re-discuss where we stand and update here. We did some initial research on the issue but haven't gotten to dive into it more yet. Sorry for the inconvenience and the wait, we have a long backlog to go through.

**Comment by thopiekar** [comment]
I'm experiencing the same issue here. I think this bug breaks one of the biggest advantages of library panels, which allows reusing panels across dashboards when they are designed to use repeating rows.

@natellium Are there any news regarding this ticket? 🙂

**Comment by MSDkuznetsov** [comment]
Hey, we have the same issue, it's quite a problem since the design is using Library Panels A LOT and they are very broken in so many areas :( This is the biggest benefit of Grafana but it doesn't work !

**Comment by Rumbelstilzchen** [comment]
It seems to be solved in 10.1.1
Can someone please confirm? 
So far I don't dare to upgrade my productive instance.

**Comment by ivanortegaalba** [comment]
@Rumbelstilzchen I was trying to reproduce it in play.grafana.com, which runs version 10.2 and it seems to be fixed

https://play.grafana.org/d/PNF4_aKVk/replicating-23-59805?orgId=1&var-query0=0&var-query0=1

**Comment by ivanortegaalba** [comment]
@oliverrahner I'm trying [your shared dashboard](https://play.grafana.org/d/PNF4_aKVk/replicating-23-59805?orgId=1&var-query0=0&var-query0=1), and now it seems to work, after some changes in the repeating code. Can you try to reproduce it in the latest version of Grafana?

**Comment by oliverrahner** [comment]
It's been a long time, the actual use case I had needs some rework so that I can use it again. But at least I can confirm the same that you saw, my minimum example seems fixed.

**Comment by ivanortegaalba** [comment]
Closing the issue then, since this use case seems to be solved, thanks for confirming 🙏

**Comment by mhoran** [comment]
Confirmed fixed across my multiple dashboards with library panels on Grafana Cloud (10.2.0). Thanks!

**Comment by mhoran** [comment]
Repeating library panels are broken again for me in the latest Grafana Cloud: #78368.
